### DISCUSSION TITLE
16-bit Windows applications and pointers

### DISCUSSION THREAD

**Discussion by Wall-AF** [NationalSecurityAgency/ghidra#2584] [Q&A]
Over the last few months I've been climbing a mountain (a nanometre at a time), but I’m wondering if the configuration in `/Processors x86/data/languages/x86-16.cspec` describing `<pointer_size value="2" />` is correct.  I’ve noticed that actually all pointers are 32-bit comprising of either an implicit or explicit `segment` followed by the `offset`.  The implicit `segment` is defined in the assembly instruction (as one of `DS`, `SS` or `ES`).

Therefore, is the definition wrong???

**Comment by ghidra1** [comment]
The definition is incomplete but is limited by GHIDRA's current capabilities.  

Ideally, GHIDRA needs to support multiple pointer sizes and types for a specific compiler spec, however the current GHIDRA Data Organization only supports the notion of a single pointer size and is generally assumed to refer to the default address space in many cases.  There are a number of other pointer attributes which would ideally be supported as well, including the notion of address space a pointer variable references.  In addition, size alone is insufficient to handle proper de-referencing in many cases (e.g., offset pointers, bit-shifted pointers, etc.).

**Reply by Wall-AF** [reply]
Is there anything in the pipeline to support this?

**Comment by ghidra1** [comment]
Nothing at this point.  It's on the wish list.

**Reply by Wall-AF** [reply]
Is there some technical documentation (not generally available) that I could gain access to that would enable me to better understand Ghidra's internal design so that I could help in this quest???

So far, I've been stepping through the Java code in order to make the changes I've already made for myself!

**Comment by ghidra1** [comment]
I have not researched a solution for your specific use-case involving segmented addressing issues, but do realize that many architectures support the notion of multiple pointer sizes and characteristics which can influence pointer interpretation and de-referencing.  The hope is the architectural changes we are currently considering will improve the "datatype attribute" storage capabilities to provide sufficient flexibility to more easily tackle new use cases such as yours without the need to incrementally change the storage schema and low-level API.  Such changes always have a big impact on data migration, datatype archives, multi-user conflict detection/merge and diff.   With such a capability in place we can more easily introduce new datatype attributes, although it is still likely that a significant retrofit effort would be required when introducing new attribute which must be handled by the GUI and analysis chain (e.g., decompiler).

I apologize for our lack of design documentation and a more immediate solution to your issue.  Your willingness to assist is appreciated.  We encourage your experimentation with solving your specific use case and passing along anything you find helpful.  If you find a solution which does not entail new pointer attributes it may be possible to work them in sooner than later.

**Reply by Wall-AF** [reply]
I know you're busy, but has there been _any_ progress???

**Reply by ghidra1** [reply]
No changes are in the wind for Data Organization at this point.  It is worth noting that you should be able to create 32-bit pointers when required.  Recent corrections have been made to the interpretation for in-memory 32-bit pointers for 16-bit code where segment/offset interpretation was incorrect. (see [commit](https://github.com/NationalSecurityAgency/ghidra/commit/9a0f630f38126950e371718a9ae67912574c726e) )

**Reply by Wall-AF** [reply]
Yes, that commit helps.  However if you have a structure and a pointer to that structure, whilst it has no size override you get nice(ish) decompilation.  Unfortunately, once you override the size, all goes wierd!

**Reply by Wall-AF** [reply]
Here's an example:

(I know Ghidra understands a `this` pointer by using the calling convention `__thiscall`, but using 16-bit Windows, I need one for `cdecl` and one for `pascal` (left-to-right param stacking) so I've made it just an ordinary param for now!)

In the output below, I have specified the type of `pNewClass` to be `CRuntimeClass *32` (a 32-bit pointer)
```
AFX_CLASSINIT * __cdecl16far
AFX_CLASSINIT::AFX_CLASSINIT(AFX_CLASSINIT *this,CRuntimeClass *pNewClass)
{
    LPVOID pvVar1;
    
    if (this == NULL) {
        pvVar1 = operator_~_new(1);
        this = (AFX_CLASSINIT *)pvVar1;
        if (this == NULL) {
            return NULL;
        }
    }
    *(undefined4 *)((int)pNewClass + 0x10) = CRuntimeClass::pFirstClass;
    CRuntimeClass::pFirstClass = pNewClass;
    return this;
}
```

In the output below, I have specified the type of `pNewClass` to be `CRuntimeClass *` (a 16-bit pointer)
```
AFX_CLASSINIT * __cdecl16far
AFX_CLASSINIT::AFX_CLASSINIT(AFX_CLASSINIT *this,CRuntimeClass *pNewClass)
{
    LPVOID pvVar1;
    undefined2 in_stack_00000008;
    
    if (this == NULL) {
        pvVar1 = operator_~_new(1);
        this = (AFX_CLASSINIT *)pvVar1;
        if (this == NULL) {
            return NULL;
        }
    }
    pNewClass->m_pNextClass = CRuntimeClass::pFirstClass;
    CRuntimeClass::pFirstClass = (CRuntimeClass *)_pNewClass;
    return this;
}
```

I would appreciate if someone would give me some idea where to look in the C/C++ decompiler code that is causing these anomalies because I'm struggling to find it!!!

Thanks.

**Reply by ghidra1** [reply]
Our hope is to switch over to a function attribute someday so other calling conventions can adopt the behavior of an instance method with a `this` parameter.  Unfortunately, this too I can't provide a timeline, but we are aware of this drawback.  For now, you will need to rely on explicit signatures with the this parameter specified (which will differ from the formal declaration).

**Reply by Wall-AF** [reply]
Thanks for that.  Meanwhile, can you give me a tip as to where to look for the functionality that attempts to identify elements within structures, as I believe that may be the source of the issues?

I.e., the different interpretations of:
`    *(undefined4 *)((int)pNewClass + 0x10) = CRuntimeClass::pFirstClass;`
versus
`    pNewClass->m_pNextClass = CRuntimeClass::pFirstClass;`

**Reply by ghidra1** [reply]
This is baked into Decompiler C++ code based upon the definition of the structure definition associated with the pointer pNewClass.  There is a popup menu action in the Decompiler which can attempt fill-out a structure definition (e.g., pNewClass) based upon such references.  Right-click on Structure pointer declaration for pNewClass and select `Auto Fill in Structure`

**Reply by Wall-AF** [reply]
Where is that option?  I've tried the 3 views `Data Type Manager`, `Decompile` and the `Listing`!

FYI the interpretation example I gave previously came from the `Decompile` view.

**Reply by ghidra1** [reply]
The action is in the `Decompile` view on the context-sensitive popup menu when right-clicking on an appropriate pointer declaration

**Reply by Wall-AF** [reply]
This is all I see:
![image](https://user-images.githubusercontent.com/67698927/115030962-9030ee00-9ebf-11eb-943a-b58d383deb9a.png)

And there's nothing like it in any of the 4 sub-menus.

**Reply by Wall-AF** [reply]
See https://github.com/NationalSecurityAgency/ghidra/discussions/2899 for another example.

**Reply by ghidra1** [reply]
The popup context (mouse location) must be pointer - I can't tell from your image since it is obscured by popup menu

**Reply by Wall-AF** [reply]
I tried both locations on the function definition line; on the type `CRuntimeClass *`, on the variable name `pNewClass`.  Neither of these worked, or triggered the function `adjustCreateStructureMenuText` in class `CreateStructureVariableAction`, however it was triggered when I tried over the local variable declaration `LPVOID pvVar1` but still no menu item appeared containing any of these options: `Auto Create Structure`, `Auto Fill in Class Structure`, `Auto Fill in Structure` or `Auto Create Class Structure`.  I also tried on other locations where `pNewClass` was shown.

I think getting that to work might not solve my issue though!  I've been spending time looking through the C/C++ (especially the `Action...` classes) to find somewhere that interprets a pointer to a `struct` with an additional non zero offset (as in ` MOV (d)word ptr ES:[BX + offset], value`), but I'm not having much luck!  That's where I was hoping to find knowledge...

**Reply by ghidra1** [reply]
Without a Ghidra project file to troubleshoot (*.gzf) I don't think I can provide much more assistance on this.

**Reply by Wall-AF** [reply]
How do I produce one?

**Reply by ghidra1** [reply]
Export action from File menu

**Reply by Wall-AF** [reply]
Here's one that I've zipped to store!
[VOICEBAR.EXE.gzf.zip](https://github.com/NationalSecurityAgency/ghidra/files/6339247/VOICEBAR.EXE.gzf.zip)
Hope this helps.

**Reply by ghidra1** [reply]
It appears the root cause of your issue is the default pointer size of 2-bytes vs the pointer size you are attempting to trigger on (i.e., 4-bytes).

If you are running from Eclipse from a source pull, you comment-out the following lines in `DecompilerStructureVariableAction.java` (~line 68)  and similar lines in `ListingStructureVariableAction.java` (~line 80) - although the structure does not fill-out as expected.  A non-default pointer size is likely causing an issue in the implementation.

**Reply by Wall-AF** [reply]
Superb.  That works.  Except, you end up with a 16-bit pointer which messes up the stack!!!

**Reply by Wall-AF** [reply]
In class `FillOutStructureCmd` method `applyTo` (~line 155), I found a new pointer is created.  When I modify the creation to include the current size, the calculated references reappear.  I'm back to square 1!  It calls the decompiler code to create the syntax!

**Reply by ghidra1** [reply]
I am hoping @emteere or @caheckman can provide additional input/suggestion on this

**Comment by Wall-AF** [comment]
[DecompilerDebugOutput_16-bit_vs_32-bit_pointers.zip](https://github.com/NationalSecurityAgency/ghidra/files/6502698/DecompilerDebugOutput_16-bit_vs_32-bit_pointers.zip)
Above is a very simple example illustrating the issue, taken from the Decompiler's Debug output.  @emteere or @caheckman do you have any suggestions please?

I am currently littering the C++ code with `cout`'s in the vain hope of determining where the decompiler's process varies between the 2 examples, but it is proving mesmerisingly difficult!

**Reply by Wall-AF** [reply]
From what I'm seeing, it's as if the decompiler is able to effectively ignore the `segment` part of any address thus allowing _near_ (16-bit) pointers, but when given an explicit _far_ (32-bit) one the `segment` and (in my case) the (BX) register are treated as individual entities and not joined to make a 32-bit pointer.

How can this be rectified???

**Reply by Wall-AF** [reply]
Similarly, if you create a `struct` with just a 16-bit pointer to your object and use that as your parameter spec, that dereferences fine (leaving unresolved stack references).  Adding anything to that `struct` (the appropriate size - 16 bits) to fix the stack alignment breaks the dereferencing!

I think a new `Rule` maybe needed, but I'm struggling to know exactly how and what to add!

**Reply by NancySadkov** [reply]
I think this asks for a general framework. Consider the int21:AH syscalls, That is basically the same as segmented calls, but with int instruction picking the segment. Ideally having say a DOS memory dump, you want to be able to jump into the syscall code without navigating through the dispatcher. Then both the analyzer, the scripts and the users could use it.
